azurelinuxagent/ga/remoteaccess.py

```diff
@@ -0,0 +1,163 @@
+                    self.remote_access = self.protocol.client.remote_access
+                    self.handle_remote_access()
+            else:
+                logger.info("Non JIT enabled client.")
+        except Exception as e:
+            msg = u"Exception processing remote access handler: {0}".format(
+                ustr(e))
+            logger.warn(msg)
+            logger.info(str(e))
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.HandleRemoteAccess,
+                      is_success=False,
+                      message=msg)
+            return     
+
+    def handle_remote_access(self):
+        logger.verbose("Entered handle_remote_access")        
+        if self.remote_access is not None:
+            # Get JIT user accounts.
+            all_users = self.os_util.get_users()
+            jit_users = set()
+            for usr in all_users:
+                if usr[4] == REMOTE_ACCESS_ACCOUNT_COMMENT:
+                    jit_users.add(usr[0])     
+            for acc in self.remote_access.user_list.users:
+                raw_expiration = acc.expiration
+                account_expiration = datetime.strptime(raw_expiration, REMOTE_USR_EXPIRATION_FORMAT)
+                now = datetime.utcnow()
+                if acc.name not in jit_users and now < account_expiration:
+                    self.add_user(acc.name, acc.encrypted_password, account_expiration)
+
+    def add_user(self, username, encrypted_password, account_expiration, retry=MAX_TRY_ATTEMPT, throttle=FAILED_ATTEMPT_THROTTLE):
+        created = False
+        attempt_count = 0
+        while attempt_count < retry:
+            try:
+                attempt_count += 1
+                expiration_date = account_expiration + timedelta(days=1) - datetime.utcnow()
+                logger.verbose("add_user: Adding user {0} with expiration in {1} day(s)".format(username, expiration_date.days))
+                self.os_util.useradd(username, expiration_date.days, REMOTE_ACCESS_ACCOUNT_COMMENT)
+                created = True
+                cache_file = os.path.join(conf.get_lib_dir(), "remote_access_pwd.dat")
+                prv_key = os.path.join(conf.get_lib_dir(), TRANSPORT_PRIVATE_CERT)
+                pwd = self.cryptUtil.decrypt_secret(encrypted_password, prv_key, cache_file, None)
+                self.os_util.chpasswd(username, pwd, conf.get_password_cryptid(), conf.get_password_crypt_salt_len())
+                self.os_util.conf_sudoer(username)
+                logger.info("add_user: User '{0}' added successfully with expiration in {1} day(s)".format(username, expiration_date.days))
+                return attempt_count
+            except OSError as oe:
+                self.handle_failed_create(username, oe.strerror, created)
+            except Exception as e:
+                self.handle_failed_create(username, str(e), created)
+            time.sleep(throttle)
```

Comment:
Why would sleeping up to 5 seconds make this more robust? Or more precisely, why do you expect to have to retry?

Reply:
... I was thinking in case there was a transient error here (thinking a network blip talking with the wire server for example) and throttling in a tight retry loop is always a good practice so as not to throttle a CPU and allow transient errors to pass.... that said, I'm no longer making any calls that could result in transient errors within reason... No need for this here, I'll remove it.

---
In reply to: [192196799](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192196799) [](ancestors = 192196799)

Reply:
I don't see any network communication in this method.  I don't think the retry is necessary either, or rather I would like to see data that it is in fact needed.

---
In reply to: [192201305](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192201305) [](ancestors = 192201305,192196799)